renderdoc/core/resource_manager.cpp

```diff
@@ -51,44 +51,59 @@ void SetReplayResourceIDs()
 
 INSTANTIATE_SERIALISE_TYPE(ResourceManagerInternal::WrittenRecord);
 
-bool MarkReferenced(std::map<ResourceId, FrameRefType> &refs, ResourceId id, FrameRefType refType)
+FrameRefType ComposeFrameRefs(FrameRefType first, FrameRefType second)
 {
-  if(refs.find(id) == refs.end())
+  if(second != eFrameRef_None)
   {
-    if(refType == eFrameRef_Read)
-      refs[id] = eFrameRef_ReadOnly;
-    else if(refType == eFrameRef_Write)
-      refs[id] = eFrameRef_ReadAndWrite;
-    else    // unknown or existing state
-      refs[id] = refType;
-
-    return true;
-  }
-  else
-  {
-    if(refType == eFrameRef_Unknown)
-    {
-      // nothing
-    }
-    else if(refType == eFrameRef_ReadBeforeWrite)
-    {
-      // special case, explicitly set to ReadBeforeWrite for when
-      // we know that this use will likely be a partial-write
-      refs[id] = eFrameRef_ReadBeforeWrite;
-    }
-    else if(refs[id] == eFrameRef_Unknown)
+    switch(first)
     {
-      if(refType == eFrameRef_Read || refType == eFrameRef_ReadOnly)
-        refs[id] = eFrameRef_ReadOnly;
-      else
-        refs[id] = eFrameRef_ReadAndWrite;
-    }
-    else if(refs[id] == eFrameRef_ReadOnly && refType == eFrameRef_Write)
-    {
-      refs[id] = eFrameRef_ReadBeforeWrite;
+      case eFrameRef_None:
+      case eFrameRef_Write: return second;
+      case eFrameRef_Read:
+        if(second != eFrameRef_Read)
+          return eFrameRef_ReadBeforeWrite;
+        break;
     }
   }
+  return first;
+}
 
+FrameRefType ComposeFrameRefsUnordered(FrameRefType first, FrameRefType second)
+{
+  if(first == eFrameRef_None)
+    return second;
+  if(second == eFrameRef_None)
+    return first;
+  if(first == eFrameRef_ReadBeforeWrite || second == eFrameRef_ReadBeforeWrite)
+    // Conservatively assume the ReadBeforeWrite occurs before the other ref
+    return eFrameRef_ReadBeforeWrite;
+  if((first == eFrameRef_Read) ^ (second == eFrameRef_Read))
```

Comment:
I think `if((first == eFrameRef_Read) != (second == eFrameRef_Read))` should generate identical assembly and IMO is a bit clearer about the intention here.